[IA-1776] Implement runtime PATCH route

rtitle opened the pull request:
https://broadworkbench.atlassian.net/browse/IA-1776

This adds `PATCH /api/google/v1/runtimes/{project}/{name}` with equivalent functionality to `PATCH /api/cluster/{project}/{name}`.

I also made a couple of improvements/modifications in the new version:
1. Made dedicated models for `PATCH` requests, instead of overloading `RuntimeRequest`
2. Moved more work to Back Leo. Now `RuntimeServiceInterp` (front Leo) only does request validation and DB updates, and the actual GCP updates are done by the PubSub message subscriber (back Leo).

TODO:
- Fill in `LeoPubsubMessageSubscriberSpec`
- Swagger
- fiab testing

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

rtitle committed:
Structure for PATCH in runtime routes

rtitle committed:
Implement RuntimeServiceInpterp changes

rtitle committed:
Fix bugs and tests

rtitle committed:
Add some comments

rtitle committed:
Add update request decoders

rtitle committed:
Add TODOs for intended tests

rtitle committed:
Add HttpRoutesSpec

rtitle committed:
Update RuntimeServiceInterpSpec

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -354,6 +406,122 @@ class RuntimeServiceInterp[F[_]: Parallel](blocker: Blocker,
+   * and potentially sends a PubSub message (or throws an error). The PubSub receiver does not
+   * do validation; it is assumed all validation happens here.
+   */
+  private[service] def processUpdateRuntimeConfigRequest(req: UpdateRuntimeConfigRequest,
```
I think this logic is the same as in `LeonardoService`. I added a lot of unit tests to make sure.

kyuksel replied:
The method is quite long. Does it make sense to factor out the `case`s so it's easier to follow at a glance? It also may (or may not) help factoring out some lines that are duplicated.

rtitle replied:
Yeah, I'll see if anything makes sense to factor out. However having a lot of separate methods is not (to me) always easier to follow if the logic is complex.

Qi77Qi replied:
yea. I don't like looking at code with too many unnecessary small functions where I have to jump around a lot

kyuksel replied:
I find it easier to look at "table of contents" before diving into the "doc" but I see yous' perspective as well  ☺️

rtitle committed:
Added case to LeoPubsubMessageSubscriberSpec

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -201,13 +210,30 @@ class RuntimeRoutes(runtimeService: RuntimeService[IO], userInfoDirectives: User
       _ <- runtimeService.startRuntime(userInfo, googleProject, runtimeName)
     } yield StatusCodes.Accepted
+
+  private[api] def updateRuntimeHandler(userInfo: UserInfo,
```
I wonder if it makes sense to put `UserInfo`, `GoogleProject`, `RuntimeName` in `RuntimeServiceContext`

rtitle replied:
Not all routes have `GoogleProject` and `RuntimeName` (e.g. list clusters)… also I feel like `UserInfo` is better left explicit.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -201,13 +210,30 @@ class RuntimeRoutes(runtimeService: RuntimeService[IO], userInfoDirectives: User
+  implicit val gceConfigDecoder: Decoder[RuntimeConfigRequest.GceConfig] = Decoder.instance { x =>
+    for {
+      machineType <- x.downField("machineType").as[Option[MachineTypeName]]
+      diskSize <- x
```
if we change `diskSize` type to dedicated type(think there's one already?), we don't have to do the negative check here.

rtitle replied:
Yeah that's a good idea
update: I did this

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -254,6 +280,56 @@ object RuntimeRoutes {
+    for {
+      machineType <- x.downField("updatedMachineType").as[Option[MachineTypeName]]
+      diskSize <- x
+        .downField("updatedDiskSize")
```
same comment

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -254,6 +280,56 @@ object RuntimeRoutes {
+      rc <- x.downField("updatedRuntimeConfig").as[Option[UpdateRuntimeConfigRequest]]
+      as <- x.downField("allowStop").as[Option[Boolean]]
+      ap <- x.downField("updatedAutopause").as[Option[Boolean]]
+      at <- x.downField("updatedAutopauseThreshold").as[Option[Int]]
```
use  `Option[FiniteDuration]` for `updatedAutopauseThreshold`

rtitle replied:
I left this as `Int` because we interpret it as `minutes` and convert to `FiniteDuration` in the yield block.

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala:
```diff
@@ -108,12 +114,21 @@ object LeoPubsubMessage {
-      with Serializable
+  final case class UpdateRuntimeMessage(runtimeId: Long,
+                                        newMachineType: Option[MachineTypeName],
+                                        stopUpdateMachineType: Boolean,
```
what does `stopUpdateMachineType` mean? maybe add a comment?

rtitle replied:
I can add a comment.. if `stopUpdateMachineType == true` then the runtime should go through a stop-update transition (via `FollowupDetails`). If `false` then the machine type can be changed immediately.

(This also implies that there is no more use for the `StopUpdate` message type, as `UpdateRuntimeMessage` will handle the stop-update flow.)

Qi77Qi replied:
shall we delete `StopUpdate` message then?

rtitle replied:
It's still used by `LeonardoService`.. I really want to delete `LeoRoutes` and `LeonardoService` ;)

Qi77Qi replied:
I'm thinking once we migrate terra-ui to use new route, we can at least delete the patch implementation in leonardoservice since patch is added recently and mostly likely no one is using it

kyuksel replied:
I wonder if `stopBeforeUpdatingMachineType` or `stopToUpdateMachineType` may be more explanatory without before overly verbose  🤔

rtitle replied:
I like `stopToUpdateMachineType`, I can change it

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -281,6 +293,46 @@ class RuntimeServiceInterp[F[_]: Parallel](blocker: Blocker,
+      runtimeOpt <- clusterQuery.getActiveClusterByNameMinimal(googleProject, runtimeName).transaction
+      runtime <- runtimeOpt.fold(F.raiseError[Runtime](RuntimeNotFoundException(googleProject, runtimeName)))(F.pure)
+      // throw 404 if no GetClusterStatus permission
+      hasPermission <- authProvider.hasNotebookClusterPermission(runtime.internalId,
```
should this be 403?

rtitle replied:
Our pattern is to 404 (e.g. pretend the runtime doesn't exist) if the user doesn't have `GetClusterStatus` permission. We return 403 if the user can _view_ the runtime but can't perform some other action.

Qi77Qi replied:
gotcha, that's very helpful...can you add the comment in code as well :)

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -363,3 +439,25 @@ final case class CreateRuntime2Request(labels: LabelMap,
+}
+
+final case class UpdateRuntimeRequest(updatedRuntimeConfig: Option[UpdateRuntimeConfigRequest],
+                                      allowStop: Option[Boolean],
```
can we make this `Boolean` instead of `Option[Boolean]`? We're defaulting it to `false` anyway, and having it as `Boolean` makes it a cleaner on `caller` side (we don't have to do `.getOrElse(false)`` everywhere

Qi77Qi commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -354,6 +406,122 @@ class RuntimeServiceInterp[F[_]: Parallel](blocker: Blocker,
+                                                      reqNumPreemptibles)) =>
+        for {
+          // should num workers be updated?
+          updatedNumWorkers <- reqNumOfWorkers.flatTraverse[F, Int] {
```
minor rename `updated` to `target`? `updated` reads to me as if the update has already happened

Qi77Qi reviewed (commented):
a few minor comments..
Looks great! so much better than the old patch endpoint. Very glad to see the changes :)

rtitle commented:
Thanks for the review @Qi77Qi! I'll apply those comments (am also updating Swagger)

rtitle committed:
PR feedback

rtitle committed:
Attempt to update swagger, haven't tested yet

rtitle committed:
Fix unit test

kyuksel commented on http/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -316,6 +316,57 @@ paths:
+          required: true
+          schema:
+            type: string
+        requestBody:
```
I get an indentation error for the above line when I copy/paste the `yaml` content onto http://editor.swagger.io.

rtitle replied:
should be fixed now

kyuksel commented on http/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -1849,6 +1900,74 @@ components:
         customClusterEnvironmentVariables:
           type: object
           description: Optional environment variables to be set on the cluster.
+    UpdateRuntimeConfig:
```
add `description` here too?

rtitle replied:
done

kyuksel commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:
```diff
@@ -269,6 +271,46 @@ class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Concurrent](
+          runtimeConfig.cloudService.interpreter.updateMachineType(UpdateMachineTypeParams(runtime, m, now))
+        }
+      }
+    } yield ()
```
Do we need to add/update `isUpdating` flag?

rtitle replied:
Yeah… that PR isn't merged yet, will probably need to coordinate with @akarukappadath about who goes first. (I'm happy to rebase this PR and incorporate that flag.)

rtitle committed:
Swagger fixes

kyuksel commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterpSpec.scala:
```diff
@@ -359,4 +361,262 @@ class RuntimeServiceInterpSpec extends FlatSpec with LeonardoTestSuite with Test
+    res.unsafeRunSync()
+  }
+
+  it should "update a GCE machine type in Running state" in {
```
Would it be good to test states other than `Running` and `Stopped`?

rtitle replied:
Like Starting, Stopping, Creating, Deleting? Hm maybe we should disallow updates in those states (and test for it).

kyuksel replied:
makes sense to me

kyuksel reviewed (commented):
Looks good overall. Questions/comments inline.

rtitle commented:
@Qi77Qi will double check on the Swagger stuff. But here is my thinking:
1. I saw that some route definitions had `security:` and some didn't. I think if it's not present it just uses the security definition at the top level, and can be specified per-route for overrides. So I just removed all of them where it's not needed.
2. I noticed that Swagger UI wasn't displaying all the fields for `runtimeConfig`. Based on https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/ I changed it to:
```
updatedRuntimeConfig:
  oneOf:
    - $ref: "#/components/schemas/UpdateGceConfig"
    - $ref: "#/components/schemas/UpdateDataprocConfig"
  discriminator:
    propertyName: cloudService
```
That seems to make it show up in the UI properly.

rtitle committed:
PR feedback and some bugfixes

kyuksel reviewed (approved):
Thanks for the refactoring, status checking and the test additions 👍

Qi77Qi commented on http/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -174,17 +169,12 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/ErrorReport"
-      security:
```
are these not needed?

rtitle replied:
I think it inherits from the top level: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/swagger/api-docs.yaml#L22

Per-route security settings are only needed for overrides.

Looking on dev the swagger routes are authenticated correctly: https://leonardo.dsde-dev.broadinstitute.org/#/runtimes/listRuntimesByProject
